### ISSUE TITLE
BUG: scipy.linalg.expm generates a large amount of sys_enter_sched_yields on linux systems

### ISSUE THREAD

**Issue by amcneil98** [scipy/scipy#17919]
### Describe your issue.

When running a script containing frequent calls to scipy.linalg.expm I noticed a huge performance disparity (>10x) between machine A running MacOS on a M1 chip and machine B running an x86 CPU on Linux.  This error is confirmed present on the following distributions: Debian 11, CentOS 7, and on both Intel, and AMD CPUs.  This issue is evidenced by high CPU utilization going to the system space irrespective of number of CPU threads present to the system.  Debugging the code example by running the following command: `sudo perf stat -e 'syscalls:sys_enter_*' python minimal.py` reveals that the minimal example below generates around 80 million `sys_enter_sched_yields` which result in the high CPU usage.  The jax implementation (`jax.scipy.linalg.expm`) of this function only generates around 2 million `sys_enter_sched_yields`.  With this issue, code containing many calls to `expm` is extremely slow when parallelized and results in abnormally high cpu usage on Linux.

### Reproducing Code Example

```python
from scipy.linalg import expm
import numpy as np

a = abs(np.random.randn(15, 15)) + 1
num = 100000
i = 0
while i < num:
    expm(a)
    i += 1
```

### Error message

```shell
None
```

### SciPy/NumPy/Python version information

```blas_mkl_info:   NOT AVAILABLE blis_info:   NOT AVAILABLE openblas_info:     libraries = ['openblas', 'openblas']     library_dirs = ['/usr/local/lib']     language = c     define_macros = [('HAVE_CBLAS', None)]     runtime_library_dirs = ['/usr/local/lib'] blas_opt_info:     libraries = ['openblas', 'openblas']     library_dirs = ['/usr/local/lib']     language = c     define_macros = [('HAVE_CBLAS', None)]     runtime_library_dirs = ['/usr/local/lib'] lapack_mkl_info:   NOT AVAILABLE openblas_lapack_info:     libraries = ['openblas', 'openblas']     library_dirs = ['/usr/local/lib']     language = c     define_macros = [('HAVE_CBLAS', None)]     runtime_library_dirs = ['/usr/local/lib'] lapack_opt_info:     libraries = ['openblas', 'openblas']     library_dirs = ['/usr/local/lib']     language = c     define_macros = [('HAVE_CBLAS', None)]     runtime_library_dirs = ['/usr/local/lib'] Supported SIMD extensions in this NumPy install:     baseline = SSE,SSE2,SSE3     found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2,AVX512F,AVX512CD,AVX512_SKX,AVX512_CLX     not found = AVX512_KNL,AVX512_KNM,AVX512_CNL,AVX512_ICL```

**Comment by ilayn** [comment]
I am not very knowledgeable about the Linux systems or the nature of `sys_enter_sched_yields` but this is expected when you want to parallelize functions that are not meant for running in parallel. I would guess that the reason for it is the Cythonized code is doing something that is not meant for the system that is not performing (didn't catch which one but I'd say M1 is the slower one if I have to guess). 

Not sure if it applies to your case but expm accepts nD arrays. So maybe that would make things easier?

**Comment by amcneil98** [comment]
Sorry for the delay.

I do not believe this is expected behavior.  I would expect a single call to `expm` to potentially break GLI based on what is going on under the hood, and compute the result across multiple cores.  This would be uniform behavior on both my M1 machine and x86 machine.  However, this is not what I see.  On the M1 machine `expm` does not seem to break GLI and stays confined to 1 core.  On the x86 machine it breaks GLI, system CPU usage reaches 100%, and the system will spend all of its CPU resources to satisfy the `sys_enter_sched_yeilds`.  The end result is that the M1 machine can complete the call faster and when paralleled with the `multiprocessing` library.

>Not sure if it applies to your case but expm accepts nD arrays. So maybe that would make things easier?
Unfortunately since my code is for a time domain simulation I cannot parallelize my calls to `expm`.

> Oh and do these sys_enter_sched_yields stay dorment or they slowly dissipate after the code execution is finished?

They dissipate.

**Comment by datapythonista** [comment]
@amcneil98, maybe you're right about the scheduler system call, but I don't think the number of calls is necessarily the problem, even if the jax version uses much less. Also, even if your assumption is right, I don't think you will be able to make comparisons in both systems in terms of system calls, as I guess you can't really see what's going on in the Mac.

What I'd personally try is to use a profiler instead, and see where most of the time is spent in both implementations. Maybe there is a function that takes a lot of time in Linux and not in Mac, and that gives us a clue on where the difference in time is coming from. Did you try that? What do you think?

**Comment by amcneil98** [comment]
@datapythonista The profiler says that the majority of the time is spent in `_matfuncs.py:213(expm)` (the scipy expm function) with total time of 60 seconds on the M1 and 135 seconds on the Linux machine.  However, I had previously been running an outdated scipy (1.9.1) on the M1 machine.  Running the profiler with scipy 1.9.1 reduced the time spent in in the expm function on the M1 down to 11 seconds, however the Linux machine remained the same.  Additionally there are a quite a few items in the profiler from the Linux machine that take several orders of magnitude longer to complete. Here are outputs of `python -m profile -s time minimal.py` for the [M1 machine](https://github.com/scipy/scipy/files/10680545/minimal_m1_output.txt) and [Linux machine](https://github.com/scipy/scipy/files/10680546/minimal_linux_output.txt).  Both outputs are on scipy 1.10.0 and numpy 1.24.2.

**Comment by ilayn** [comment]
From this distance to the code, I don't know what else I can help with since I have no idea what is happening in your system. The parallelization can only help if you do multiple expm calls simultaneously (maybe) but the computation of expm cannot be parallelized. The algorithm is not suitable for it. If your LAPACK/BLAS is multithreaded then you might have a better performance (again maybe). 

> I do not believe this is expected behavior. I would expect a single call to expm to potentially break GLI based on what is going on under the hood, and compute the result across multiple cores

That depends on your install and the underlying LAPACK/BLAS configuration. If you have the stock pip install, it comes with a single threaded OpenBLAS and that won't let you do what I think you are describing. You might really benefit from building SciPy yourself with a multithreaded BLAS. Conda is also another option as it comes with Intel MKL which might be ready out of the box.

> Unfortunately since my code is for a time domain simulation I cannot parallelize my calls to expm.

Matrix exponentials are almost always the wrong tool for time domain simulations. Consider using a time domain solver instead by having a time step increments. Otherwise you will have numerical problems and blowups. If it is a linear system a state space simulation is a much better choice (scipy.signal, python-control, harold and so on are all capable of it) .

**Comment by datapythonista** [comment]
@amcneil98 from the profiling you share, feels like your M1 machine is simply double as fast as your Linux machine. If with scipy 1.9.1 your code runs in 11 seconds instead of 60 in the same machine, that's something worth looking at.

So, what I understand from what you said is:

| OS | SciPy | Time |
|-|-|-|
| MacOS | 1.9 | 10 |
| MacOS | 1.10 | 60 |
| Linux | 1.9 | 120 |
| Linux | 1.10 | 120 |

My bet is that your environment with SciPy 1.9 in Mac had probably MKL installed, while you had openblas in the other environments.

I tried running your example with both openblas and mkl, here you can see the results in my (Linux) machine:

**openblas** (from conda-forge, seems to use all cores)

```
CPU times: user 23.2 s, sys: 29.8 s, total: 52.9 s
Wall time: 8.03 s
```

**mkl** (from Anaconda, seems to use only one core)

```
CPU times: user 5.11 s, sys: 16.6 ms, total: 5.13 s
Wall time: 5.12 s
```

The results are different from what you reported, but if your openblas comes from pip and is single threaded as @ilayn says, that would explain what you're reporting.

Let us know if this makes sense, and the underlying library does explain the behavior you experienced.


Btw, in case it's helpful, this is implementation of the algorithm running: https://github.com/scipy/scipy/blob/b9cf39da9c9294588cc3feb9658b94172fd7f96b/scipy/linalg/_matfuncs_expm.pyx.in#L385

**Comment by ilayn** [comment]
I also found out that Jax implementation has a quite important shortcut that avoids a lot of matrix powers. I am not sure about its validity (I might be wrong though) but that is the reason why you don't see as much activity with Jax. 

Could you please test with the following matrix 

```python
import numpy as np
import scipy.linalg as la
import jax.scipy.linalg as jla
A = np.array([[0, 1.0e-08, 0],
              [-2.006666666666667e+10,  -3.0,  2.0e+10],
              [ 6.666666666666667e+01,   0.,  -6.666666666666667e+01]])
print(la.expm(A))
print(jla(A))
```

both should give 

```python
array([[ 4.46849468e-01,  1.54044157e-09,  4.62811454e-01],
       [-5.74306778e+06, -1.52830039e-02, -4.52654271e+06],
       [ 4.47722978e-01,  1.54270485e-09,  4.63480649e-01]])
```

It is a famous example for overscaling ([link](https://blogs.mathworks.com/cleve/2012/07/23/a-balancing-act-for-the-matrix-exponential/)) and probably would be good enough to trip it. Maybe it is indeed correct to have this shortcut but I couldn't see how it can work in all cases. Otherwise we can switch to that shortcut too and get rid of many tedious steps that took me a lot of time to write and also would make me very happy to delete.

**Comment by amcneil98** [comment]
Regarding LAPACK/BLAS version, everything in all my environments is installed via pip.  As far as differences between my scipy 1.9.1 and 1.10.0 envs should have only been the scipy versions.  When I switched scipy versions with `pip install scipy==${VERSION_NUM}` the only output of the pip command was changing scipy versions regardless of platform.  

Regarding Jax.  I've found an issue with the expm command.  The issue is with the `max_squarings` argument.  The result between scipy and jax is not equivalent unless `max_squarings` is set really high in this case the same issue from scipy becomes present.  Thus I think the Jax "solution" only is a reduction in accuracy which brings about the performance boost.

@datapythonista 
The table is correct.

The difference between openblas and mlk seem to directly correspond to the issue that I'm seeing.  Testing this will take some extra time as I tend to stay away from Conda so I will have to rebuild everything in a Conda environment.

**Comment by ilayn** [comment]
Hmm that's interesting that nobody from jax complained about it apparently. The squaring computation is indeed unreasonably complicated but necessary evil and that is what is skipped in Jax implementation. For example for the example I gave above the correct squaring is 8 but jax formula and the old matlab implementation needs 31 at least. I was hoping for a silver bullet but I guess there is not one found yet.

Please let us know how it goes with MKL, highly appreciated.

**Comment by amcneil98** [comment]
Ok, I can confirm @datapythonista's findings that the scipy with MKL that is installed with `conda install -c anaconda scipy` shows a significant performance uplift on the minimal example.  The scipy installed with `conda install -c conda-forge scipy` is significantly slower than the one from pip, with 173s spent in `linalg.py`, 153s spent in `dot`, and 99s in `_matfuncs.py:expm`.  Its important to note that the conda-forge scipy is 1.8.1, where as the anaconda scipy is 1.10.0.  Attempting the following command  `conda install -c conda-forge scipy==1.10.0` installed scipy with MKL.  The MKL version from anaconda does not generate any `sys_enter_sched_yields`, where as the the conda-forge openblas version generates 1,283,977,659 `sys_enter_sched_yields`.  It seems like the difference between these two libraries is the source of the drama and the current workaround is to use the MKL version.

**Comment by ilayn** [comment]
Please also note that the OpenBLAS that comes with pip install is not fully architecture optimized. Hence it is a generic version. If you build it yourself you might also see a performance boost with OpenBLAS too. But in any case glad that things worked out for the better.

I'll ping @jakevdp just in case the expm accuracy comes up in JAX in the future.

And @martin-frbg in case there is something that can be done on the OpenBLAS side though I'm not sure what that can be.

**Comment by martin-frbg** [comment]
thanks - the issue with sched_yield has come up before, it arises from (idle) threads doing busy waits on some architectures.A direct link to perfomance has not been proven in the past except on small systems prone to overheating.
A more important factor could be that OpenBLAS uses the unoptimized reference implementation of LAPACK,
which is for the most part not parallelized and contains some known  
bottlenecks (e.g. its implementation of ZLASR)
